release-22.2: xform: use ordering from LIMIT as a hint for streaming group-by

msirek opened the pull request:
Backport 1/1 commits from #93858.

/cc @cockroachdb/release

---

Fixes #93410

A query with a grouped aggregation, a LIMIT and an ORDER BY may not always explore the best-cost query plan.

Due to the existence of unique constraints on a table, the set of grouping columns may be reduced during normalization via rule ReduceGroupingCols such that it no longer includes columns present in the ORDER BY clause. This eliminates possibly cheap plans from consideration, for example, if the input to the aggregation is a lookup join, it may be cheaper to sort the input to the lookup join on the ORDER BY columns if they overlap with the grouping columns, so that a streaming group-by with no TopK operator can be used, and a full scan of the inputs to the join is avoided.

This fix adds a new exploration rule which ensures that a grouped aggregation with a LIMIT and ORDER BY clause considers using streaming group-by with no TopK when possible.

Release note (bug fix): This patch fixes join queries involving tables with unique constraints using LIMIT, GROUP BY and ORDER BY clauses to ensure the optimizer considers streaming group-by with no TopK operation, when possible. This is often the most efficient query plan.

Release justification: low risk fix for missing exploration of streaming group-by, disabled by default

---
sql: add optimizer_use_limit_ordering_for_streaming_group_by session setting

This commit adds the `optimizer_use_limit_ordering_for_streaming_group_by`
session setting that defaults to `false`. When `true`, an exploration
rule which uses the ordering specified in a
`SELECT ... GROUP BY ... ORDER BY ... LIMIT n;` statement as an
interesting ordering to require from the input to the group-by
expression, possibly eliminating a top-k operation.  When `false`, the
exploration rule is disabled. This setting allows backport of the new
exploration rule without impacting query plans of running clusters.

Release note: None

Commit:
xform: use ordering from LIMIT as a hint for streaming group by

Fixes #93410

A query with a grouped aggregation, a LIMIT and an ORDER BY
may not always explore the best-cost query plan.

Due to the existence of unique constraints on a table, the set of
grouping columns may be reduced during normalization via rule
ReduceGroupingCols such that it no longer includes columns
present in the ORDER BY clause. This eliminates possibly cheap
plans from consideration, for example, if the input to the
aggregation is a lookup join, it may be cheaper to sort the
input to the lookup join on the ORDER BY columns if they overlap
with the grouping columns, so that a streaming group-by with no
TopK operator can be used, and a full scan of the inputs to
the join is avoided.

This fix adds a new exploration rule which ensures that a grouped
aggregation with a LIMIT and ORDER BY clause considers using
streaming group-by with no TopK when possible.

Release note (bug fix): This patch fixes join queries involving
tables with unique constraints using LIMIT, GROUP BY and ORDER BY
clauses to ensure the optimizer considers streaming group-by with
no TopK operation, when possible. This is often the most efficient
query plan.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94603)
<!-- Reviewable:end -->

Commit:
sql: add optimizer_use_limit_ordering_for_streaming_group_by session setting

This commit adds the `optimizer_use_limit_ordering_for_streaming_group_by`
session setting that defaults to `false`. When `true`, an exploration
rule which uses the ordering specified in a
`SELECT ... GROUP BY ... ORDER BY ... LIMIT n;` statement as an
interesting ordering to require from the input to the group-by
expression, possibly eliminating a top-k operation.  When `false`, the
exploration rule is disabled. This setting allows backport of the new
exploration rule without impacting query plans of running clusters.

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r1, 10 of 10 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94603#-:-NKu3KHI9lc1vqhGbJgJ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @rharding6373)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Is there a pressing motivation for backporting this rule to v22.2? I think we should probably gate it behind a session setting to avoid changing customers' query plans in a patch release.

FYI, you can backport commits from two PRs with the backport tool: `backport 12345 98765 -r 22.2` (not sure if that's what you did, but looks like you did this more manually based on the PR description).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94603#-:-NKxfnMD4pL60SLdCFk6:bezv1w4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @rharding6373)



<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> Is there a pressing motivation for backporting this rule to v22.2?

This is being backported per request from @leeeeeeeroy-jenkins for issue https://github.com/cockroachlabs/support/issues/1949
> I think we should probably gate it behind a session setting to avoid changing customers' query plans in a patch release.

It is gated behind session setting `optimizer_use_limit_ordering_for_streaming_group_by` in this PR.

> FYI, you can backport commits from two PRs with the backport tool...

Thanks

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94603#-:-NKxglMDEzKFlAqWXKSs:b9uaeti)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @rharding6373)



<!-- Sent from Reviewable.io -->

mgartner commented:
> I think we should probably gate it behind a session setting to avoid changing customers' query plans in a patch release.

I just saw #94672. Can you include that commit in this backport?





<!-- Sent from Reviewable.io -->

mgartner commented:
> It is gated behind session setting optimizer_use_limit_ordering_for_streaming_group_by in this PR.

Oh I see, sorry!





<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
I implemented the session setting first on 22.2, that's why you don't see a backport message of the 2nd commit.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94603#-:-NKxhhW33hUiunkmo3xi:b-m53pxg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @rharding6373)



<!-- Sent from Reviewable.io -->

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94603#-:-NKy3W8BD7y1SGy9Ay6Z:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @cucaroach)



<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94603#-:-NL-pzi5COMEGN5bn1sb:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @cucaroach and @DrewKimball)



<!-- Sent from Reviewable.io -->
